diff --git a/taxonomy_dupecheck.module b/taxonomy_dupecheck.module
index bee919d..2c62987 100644
--- a/taxonomy_dupecheck.module
+++ b/taxonomy_dupecheck.module
@@ -89,10 +89,12 @@ function taxonomy_dupecheck_form_taxonomy_form_term_alter(&$form, &$form_state,
   $types = variable_get('taxonomy_dupecheck_types', array());
 
   // Add the validator if we're checking terms for dupes.
-  if ($types['term']) {
-    $vids = array_filter(variable_get('taxonomy_dupecheck_vocabularies', array()));
-    if (count($vids) == 0 || in_array($form['vid']['#value'], $vids)) {
-      $form['#validate'][] = 'taxonomy_dupecheck_term_validate';
+  if (isset($types['term'])) {
+    if ($types['term']) {
+      $vids = array_filter(variable_get('taxonomy_dupecheck_vocabularies', array()));
+      if (count($vids) == 0 || in_array($form['vid']['#value'], $vids)) {
+        $form['#validate'][] = 'taxonomy_dupecheck_term_validate';
+      }
     }
   }
 
