diff --git a/smtp.mail.inc b/smtp.mail.inc
index 8a3477a..03cefb7 100644
--- a/smtp.mail.inc
+++ b/smtp.mail.inc
@@ -121,11 +121,6 @@ class SmtpMailSystem implements MailSystemInterface {
       }
       $headers['Reply-To'] = $reply;
     }
-    $properfrom = variable_get('site_mail', '');
-    if (!empty($properfrom)) {
-      $headers['From'] = $properfrom;
-      $from = $properfrom;
-    }
 
     // Blank value will let the e-mail address appear.
 
@@ -142,6 +137,8 @@ class SmtpMailSystem implements MailSystemInterface {
         }
       }
     }
+
+    $headers['From'] = $from;
     $from_comp = $this->_get_components($from);
 
     if (!valid_email_address($from_comp['email'])) {
